ntoskrnl/io/pnpmgr/pnpenum.c

```diff
@@ -0,0 +1,339 @@
+                                    RequestLink);
+
+        if (IsListEmpty(&IopPnpEnumerationRequestList))
+        {
+            break;
+        }
+
+        RemoveHeadList(&IopPnpEnumerationRequestList);
+        KeReleaseSpinLock(&IopPnPSpinLock, OldIrql);
+
+        if (Request)
+        {
+            InitializeListHead(&Request->RequestLink);
+
+            //FIXME: Check ShuttingDown
+
+            DeviceObject = Request->DeviceObject;
+            ASSERT(DeviceObject);
+
+            DeviceNode = IopGetDeviceNode(DeviceObject);
+            ASSERT(DeviceNode);
+
+            if (DeviceNode->State == DeviceNodeDeleted)
+            {
+                DPRINT("PipEnumerationWorker: DeviceNode->State == DeviceNodeDeleted\n");
+                Status = STATUS_UNSUCCESSFUL;
+                goto NextRequest;
+            }
+
+
+            DPRINT("PipEnumerationWorker: DeviceObject - %p, Request->RequestType - %X\n",
+                   DeviceObject,
+                   Request->RequestType);
+
+            switch (Request->RequestType)
+            {
+                case PipEnumDeviceOnly:
+                case PipEnumDeviceTree:
+                case PipEnumRootDevices:
+                case PipEnumSystemHiveLimitChange:
+                    DPRINT("PipEnumerationWorker: Reenumeration ...\n");
+                    Status = IopEnumerateDevice(Request->DeviceObject);
+                    IsDereferenceObject = FALSE;
+                    break;
+
+                default:
+                    DPRINT1("PipEnumerationWorker: RequestType %u is NOT IMPLEMENTED\n", Request->RequestType);
+                    Status = STATUS_NOT_IMPLEMENTED;
+                    break;
+            }
+
+NextRequest:
+            if (Request->CompletionStatus)
+            {
+                *Request->CompletionStatus = Status;
+            }
+
+            if (Request->CompletionEvent)
+            {
+                KeSetEvent(Request->CompletionEvent, IO_NO_INCREMENT, FALSE);
+            }
+
+            if (IsDereferenceObject)
+            {
+                ObDereferenceObject(Request->DeviceObject);
+            }
+
+            ExFreePoolWithTag(Request, TAG_IO);
+        }
+        else if (IsAssignResources || IsBootProcess)
+        {
+            ASSERT(FALSE);
+#if 0
+            SERVICE_LOAD_TYPE DriverLoadType = DemandLoad;
+
+            ObReferenceObject(IopRootDeviceNode->PhysicalDeviceObject);
+
+            PipProcessDevNodeTree(IopRootDeviceNode,
+                                  PnPBootDriversInitialized,
+                                  IsAssignResources,
+                                  0,
+                                  FALSE,
+                                  FALSE,
+                                  &DriverLoadType,
+                                  NULL);
+#endif
+            IsAssignResources = FALSE;
+            IsBootProcess = FALSE;
+        }
+        else
+        {
+            ASSERT(FALSE);
```

Comment:
Missing a comment to explain why this case is not possible or not currently handled...